### ISSUE TITLE
Obsidian crashes when viewing too many drawed elements on ipad and samsung tablets

### ISSUE THREAD

**Issue by devinat1** [zsviczian/obsidian-excalidraw-plugin#863]
On the latest version of excalidraw and obsidian (Obsidian 1.0.3 and excalidraw 1.7.26), obsidian completely freezes when you view an excalidraw file with too many elements inside it. 
Furthermore, when you open obsidian again, it freezes again since it is stuck on loading the excalidraw file, so the only way to use obsidian again is to quickly close the tab the excalidraw file was on.
This only happens with larger excalidraw files (around 2mb size or more).
Obsidian-excalidraw-plugin is the only add on installed.
Steps: Install fresh obsidian, create a vault, install obsidian-excalidraw-plugin, create a drawing, draw a lot, freeze, crash.

**Comment by devinat1** [comment]
I believe it is directly related to this issue on excalidraw: https://github.com/excalidraw/excalidraw/issues/5788

**Comment by slabzzz** [comment]
Yeah this is the only reason I'm not using excalidraw as my daily driver

**Comment by zsviczian** [comment]
> It happens more often on older devices, I've noticed

We are still talking about iPads right?

What do you have on the canvas? Just freedraw strokes, or other elements as well, e.g. images or embedded markdown, etc?

**Comment by zsviczian** [comment]
if you open Excalidraw.com and do the same amount of freedraw writing, do you experience any issues?

**Comment by devinat1** [comment]
Just freedraw strokes. On ipads yes, on an ipad it crashes every time, whereas on my samsung tablet it is just painfully slow.
Here is an example file causing the issue: 
[Lab-4-part-4-Drawing-2022-10-28-13.38.48.excalidraw(2).md](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/9894065/Lab-4-part-4-Drawing-2022-10-28-13.38.48.excalidraw.2.md)

**Comment by devinat1** [comment]
> if you open Excalidraw.com and do the same amount of freedraw writing, do you experience any issues?

I am not sure how to copy an md file into excalidraw.com (I tried copying the md file I sent to excalidraw.com and I can do that on windows but not sure how to do the same thing on ipad).

**Comment by zsviczian** [comment]
I copied the drawing to excalidraw.com, then clicked export to a shareable link and was able to open the drawing on the iPad.

I can't solve the performance issue in Excalidraw on my own, however, as a workaround I am thinking to add a setting to limit number of elements in zoom to fit... let me play with this a little.

In the meantime you can try to turn off zoomToFit in Excalidraw settings... though I am afraid that would be sufficient help.

**Comment by zsviczian** [comment]
Can you try this main.js (you need to unzip [main.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/9894247/main.zip)) and copy main.js to `Your_Vault/.obsidian/plugins/obsidian-excalidraw-plugin` folder.
Copying the file to this folder on the iPad is a bit more tricky since the built-in file explorer on the iPad does not show hidden folders, and `.obsidian` is a hidden folder. I use Obsidian Sync, it simply copies the plugins from the desktop to my iPad - nice and simple. On Android, you can use almost any file explorer app and it will display hidden folders without any problem.

The workaround that I implemented in this main.js will check if you have more than 1000 elements on your drawing, if you do and you are using Obsidian mobile, then it will not execute zoom to fit. This way the drawing will open without an issue... watch out, however, if you manually zoom to fit all elements, you will get into the same performance issue as before.

Please let me know if this is a workable workaround.

**Comment by devinat1** [comment]
Ok I will try that then let you know. The workaround is ok, but it is way too easy to accidentally zoom to fit too many elements.
Can it be so that if you zoom manually, you are blocked if that zoom would have too many elements in the drawing?

**Comment by zsviczian** [comment]
I think the root cause needs to be fixed. Unfortunately that will take someone a couple of days (maybe more) of hard work to figure out the issue and implement a fix.

The truth is, Excalidraw was designed as an illustration tool, not for handwritten notes. I am bullish about Obsidian and Excalidraw, however, I used to use LectureNotes on Android for taking handwritten notes in meetings, and Notability on iOS. The issue with these tools is you can only export the notes to PDF or an image format and import the image into Obsidian.

**Comment by devinat1** [comment]
Yea I wish that excalidraw was better suited for handwritten notes. I do typed notes the way excalidraw was meant to be used too, but usually do handwritten notes. I love the ability to see my handwritten notes show up in real time within my obsidian vault due to obsidian sync. There are a couple bugs and slowdowns that make using obsidian as a daily driver for handwritten notes very annoying. This one has been bugging me the most on android tablets: #874

**Comment by m-beau** [comment]
Hi there,

Just confirming that other users also experience this issue (in my case after only ~20 lines of handwritten text, leading to a 1.8MB file - not that overwhelming in theory but apparently too much for excalidraw on iOS) and that it would be worth it to fix it! Thanks a lot!

**Comment by zsviczian** [comment]
@m-beau, I think you are referring to a different problem than described in this specific issue. The issue you seem to be referring to is this: https://github.com/excalidraw/excalidraw/issues/5788

Unfortunately 5788
- is not an Obsidian plugin bug, but an Excalidraw.cor core package problem, that
- based on my testing only impacts iOS (Safari), in my personal opinion, because Apple does not fully support open web standards... and is
- very hard to debug and fix, again because Apple created such a closed environment, that prevents developers that do not use a Mac from debugging iOS web apps.

Bottom line is, yes, I know about the issue. I've already spent about 8-10 hours of my time trying to get to the bottom of it. I have a general understanding of the underlying problem, which is a Safari deficiency. Unfortunately other developers, especially those with access to a Mac, have not picked up on trying to solve this issue in the main Excalidraw thread... This is a long way of saying, I don't see a solution at hand, and can only recommend using products that support open standards. Else you can try to push this issue by adding a comment in the excalidraw.com issue list following the link above.

**Comment by devinat1** [comment]
Thanks for this. I also have a samsung tablet which has the open standard that is android, but drawing is simply slow on there. So I can choose between a fast experience on the ipad, but with the potential for crashes if I zoom out too much on accident, or a slow experience on samsung after just a few strokes, but with no crashes.

**Comment by HHChuang** [comment]
I also have similar issues in which iPad froze while opening excalidraw notes (extension is *.md). But if I change the file extension from *.md to *.excalidraw, I can use it on my iPad without crashing down. Now I use VScode or excalidraw on Safari to create a file with extension *.excalidraw and use Obsidian on iPad for handwriting. Not perfect but at least I really need the handwriting function...
